shout-button-message/src/index.ts

```diff
@@ -0,0 +1,109 @@
+    this.statusBarSummary.innerText = 'Last Shout: (None)';
+    this.node.appendChild(this.statusBarSummary);
+  }
+
+  setSummary(summary: string) {
+    this.statusBarSummary.innerText = summary;
+  }
+}
+
+/**
+ * ShoutWidget holds all the plugin's primary functionality.
+ * It also creates a widget for JupyterLab's status bar if the
+ * status bar is available.
+ */
+class ShoutWidget extends Widget {
+
+  shoutButton: HTMLElement;
+  lastShoutTime: Date | null;
+  statusBarWidget: ShoutStatusBarSummary | null;
+
+  constructor(statusBar: any) {
+    super();
+
+    // Create and add a button to this widget's root node
+    const shoutButton = document.createElement('div');
+    shoutButton.innerText = 'Press to Shout';
+    // Add a listener to "shout" when the button is clicked
+    shoutButton.addEventListener('click', this.shout.bind(this));
+    shoutButton.classList.add('jp-shout-button');
+    this.node.appendChild(shoutButton);
+    this.shoutButton = shoutButton;
+
+    // Store the last shout time for use in the status bar
+    this.lastShoutTime = null;
+
+    // Check if the status bar is available, and if so, make
+    // a status bar widget to hold some information
+    this.statusBarWidget = null;
+    if (statusBar) {
+      this.statusBarWidget = new ShoutStatusBarSummary();
+      statusBar.registerStatusItem('shoutStatusBarSummary', {item: this.statusBarWidget});
+    }
+  }
+
+  // Make an alert popup that shouts upon user click
+  shout() {
+    this.lastShoutTime = new Date();
+    window.alert('Shouting at ' + this.lastShoutTime);
+
+    // Update the status bar widget if available
+    if (this.statusBarWidget) {
+      this.statusBarWidget.setSummary('Last Shout: ' + this.lastShoutTime.toString());
+    }
+  }
+}
+
+/**
+ * JupyterLab extensions are made up of plugin(s). You can specify some
+ * information about your plugin with the properties defined here. This
+ * extension exports a single plugin, and lists the IStatusBar from
+ * JupyterLab as optional.
+ */
+const plugin: JupyterFrontEndPlugin<void> = {
+  id: 'shout_button_message:plugin',
+  description: 'An extension that adds a button and message to the right toolbar, with optional status bar widget in JupyterLab.',
+  autoStart: true,
+  // The IStatusBar is marked optional here. If it's available, it will
+  // be provided to the plugin as an argument to the activate function
+  // (shown below), and if not it will be null.
+  optional: [IStatusBar],
+  // Make sure to list any 'requires' and 'optional' features as arguments
+  // to your activate function (activate is always passed an Application,
+  // then required arguments, then optional arguments)
+  activate: (app: JupyterFrontEnd, statusBar: IStatusBar | null) => {
+    console.log('JupyterLab extension shout_button_message is activated!');
+
+    // Create a ShoutWidget and add it to the interface in the right sidebar
+    const shoutWidget: ShoutWidget = new ShoutWidget(statusBar);
```

Comment:
This is not the best architecture to achieve the feature. You should divide to conquer as much as possible and use signal to communicate between the divided nations.

So in your case a better pattern would be:

```suggestion
    const shoutWidget: ShoutWidget = new ShoutWidget();
    
    if (statusBar) {
      const statusBarWidget = new ShoutStatusBarSummary();
      
      shoutWidget.messageShouted.connect((widget: ShoutWidget, time: Date) => {
        statusBarWidget.setSummary('Last Shout: ' + this.lastShoutTime.toString());
      });
      
      statusBar.registerStatusItem('shoutStatusBarSummary', {item: this.statusBarWidget});
    }
```

This will require adding `messageShouted` to `ShoutWidget`:

```ts
class ShoutWidget extends Widget {
  // ...

  get messageShouted(): ISignal<ShoutWidget, Date> {
    return this._messageShouted;
  }
 shout() {
    this.lastShoutTime = new Date();
    window.alert('Shouting at ' + this.lastShoutTime);

    this._messageShouted.emit(this.lastShoutTime);
  }
  // ...
  private _messageShouted = new Signal<ShoutWidget, Date>(this);
}
```

Reply:
I keep running into this actually, where we try to kitchen-sink all of our documentation with every possible detail and use case, and it ends up making the docs less readable, so I am constantly advocating for simplicity, reduced complexity, and plain language, so that's where I'm coming from: Making the docs and tutorials as easy and accessible to as diverse a group of people as possible.

Reply:
I have added a note in the relevant section suggesting signals as a preferred solution in production (as noted above), and will consider this one resolved.